src/core/unit.c

```diff
@@ -4280,7 +4280,18 @@ int unit_patch_contexts(Unit *u) {
                          * UID/GID around in the file system or on IPC objects. Hence enforce a strict
                          * sandbox. */
 
-                        ec->private_tmp = true;
+                        /* Enforce private tmp directories, either via TemporaryFileSystem= or PrivateTmp=,
+                         * as they have different semantics but are both enough for sanboxing purposes. */
+                        bool tmp_found = false, var_tmp_found = false;
+                        FOREACH_ARRAY(f, ec->temporary_filesystems, ec->n_temporary_filesystems)
+                                if (path_equal(f->path, "/tmp"))
+                                        tmp_found = true;
+                                else if (PATH_IN_SET(f->path, "/var", "/var/tmp"))
+                                        var_tmp_found = true;
+
+                        if (!tmp_found || !var_tmp_found)
+                                ec->private_tmp = true;
+
```

Comment:
that said, maybe a simple fix is to not imply PrivateTmp= in case of DynamicUser=1 but instead imply TemporaryFileSystem=/tmp /var/tmp?

wouldn't that be better and simpler?

i mean, there definitely is value in keeping dynamic uids/gids local to file systems that are destroyed when the service is dead, hence changing the logic liek that would make sense to me.

Reply:
Ok I've applied this change, it's slightly backward incompatible unlike the original version but if that's what it takes